### ISSUE TITLE
90 min PPP 4.12

### ISSUE THREAD

**Issue by anish-kum** [anish-kum/PPP-Spark-2025-2026#33]
90 min PPP

**Comment by anish-kum** [comment]
<img width="332" height="476" alt="Image" src="https://github.com/user-attachments/assets/89fc1d2d-da43-41d2-ae23-cd6523edc5cd" />

What I learned:

- Training RNNs: RNNs are trained with backpropagation through time, where errors are passed across sequence steps. This is important because the model has to learn from both current and earlier inputs.

- Vanishing and exploding gradients: Gradients can shrink too much or grow too large during training. This is important because these problems make it hard for basic RNNs to learn long-term patterns.

- Long-term dependencies: Standard RNNs struggle to remember information from far back in a sequence. This is important because many sequence tasks depend on earlier context.

- Why LSTMs help: LSTMs are introduced as a better RNN variant for handling memory over longer sequences. This is important because they solve key weaknesses of simple RNNs.

2 Questions I still had:

- How exactly do LSTMs control what information to keep or forget using gates, and how are those gates trained?

- In practice, how do you tell if vanishing gradients are hurting your model versus just needing more data or tuning?
